dangerzone/updater/signatures.py

```diff
@@ -0,0 +1,642 @@
+            )
+    return signatures
+
+
+def verify_signatures(
+    signatures: List[Dict],
+    image_digest: str,
+    pubkey: Path = DEFAULT_PUBKEY_LOCATION,
+) -> None:
+    if len(signatures) < 1:
+        raise errors.SignatureVerificationError("No signatures found")
+
+    for signature in signatures:
+        # Will raise on errors
+        verify_signature(signature, image_digest, pubkey)
+
+
+def get_last_log_index() -> int:
+    SIGNATURES_PATH.mkdir(parents=True, exist_ok=True)
+    if not LAST_LOG_INDEX.exists():
+        return BUNDLED_LOG_INDEX
+
+    with open(LAST_LOG_INDEX) as f:
+        return int(f.read())
+
+
+def get_log_index_from_signatures(signatures: List[Dict]) -> int:
+    def _reducer(accumulator: int, signature: Dict) -> int:
+        try:
+            logIndex = int(signature["Bundle"]["Payload"]["logIndex"])
+        except (KeyError, ValueError):
+            log.debug("Discarding invalid logindex")
+            return accumulator
+        return max(accumulator, logIndex)
+
+    return reduce(_reducer, signatures, 0)
+
+
+def write_log_index(log_index: int) -> None:
+    last_log_index_path = SIGNATURES_PATH / "last_log_index"
+
+    with open(last_log_index_path, "w") as f:
+        f.write(str(log_index))
+
+
+def _get_images_only_manifest(input: dict) -> dict:
+    """Filter out all the non-images from a loaded manifest"""
+    output = input.copy()
+    output["manifests"] = [
+        manifest
+        for manifest in input["manifests"]
+        if manifest["annotations"].get("kind")
+        in ("dev.cosignproject.cosign/imageIndex", "dev.cosignproject.cosign/image")
+    ]
+    return output
+
+
+def _get_blob(digest: str) -> PurePath:
+    return PurePath() / "blobs" / "sha256" / digest.replace("sha256:", "")
+
+
+def _get_signature_filename(input: Dict) -> PurePath:
+    for manifest in input["manifests"]:
+        if manifest["annotations"].get("kind") == "dev.cosignproject.cosign/sigs":
+            return _get_blob(manifest["digest"])
+    raise errors.SignatureExtractionError()
+
+
+def upgrade_container_image_airgapped(
+    container_tar: Path,
+    pubkey: Path = DEFAULT_PUBKEY_LOCATION,
+    bypass_logindex: bool = False,
+) -> str:
+    """
+    Verify the given archive against its self-contained signatures, then
+    upgrade the image and retag it to the expected tag.
+
+    The logic supports both "dangerzone archives" only, which have
+    `dangerzone.json` file at the root of the tarball.
+
+    See `prepare_airgapped_archive` for more details.
+
+    :return: The loaded image name
+    """
+
+    with TemporaryDirectory() as _tempdir, tarfile.open(container_tar, "r") as archive:
+        # First, check the archive type
+        files = archive.getnames()
+        tmp_path = Path(_tempdir)
+
+        has_dangerzone_manifest = f"./{DANGERZONE_MANIFEST}" in files
+        if not has_dangerzone_manifest:
+            raise errors.InvalidImageArchive()
```

Comment:
A message about the error (either written here, or in the class description) would help for debugging purposes. For instance, I encountered this error when attempting to use an image that didn't have the `dangerzone.json` manifest, but I had to look into the code in order to understand what's the issue.

Reply:
Thanks. See 3bc463a668a7f45347d9cd43d3ecaafde4cdfc29

Reply:
Sure, here's a suggestion:

```suggestion
            raise errors.InvalidImageArchive(
                "Archives should contain a `dangerzone.json` file."
                " Consider using `dangerzone-image prepare-archive`"
                " to include it."
            )
```

My main point though is that this is the same across most of our new exception classes. None return an error message to the user, they only have it in their docstring. That's unlike the current error classes that have code like that:

```python
    def __init__(self) -> None:
        super().__init__("Input filename has not been set yet.")
```

Is it more clear now?

Reply:
D'oh! Ok, thanks!

Reply:
I did so for all error classes in 0ac916631bbe6c68f2217fe29aa4a67a041e8b70

Reply:
Well, that's a way :sweat_smile: . I think it's better to keep docstrings and error messages separated, but at this point, it's a bit minor. A few docstrings of error classes need some tweaks (e.g., `InvalidDangerzoneManifest`), but it's better than before.

We'll revisit it later on.